negotiate/negotiate.go

```diff
@@ -263,6 +263,9 @@ func NewClientContextWithFlags(cred *sspi.Credentials, targetName string, flags
 
 // Release free up resources associated with client context c.
 func (c *ClientContext) Release() error {
+	if c == nil {
```

Comment:
I do not like making Release code no-op on nil pointer.

I think it is important that user calls Release after use.

I don't see how it is possible for *ClientContext in normal user code to be nil. So if *ClientContext is nil by mistake, I want Release to panic, not to go away quietly.

Reply:
> I think it is important that user calls Release after use.

Right, I totally agree with you on this, but that assumes the creation of the resources occurs in the same spot as the release.  For example, see how OpenShift uses GSSAPI code:

GSSAPI nil safe code release: https://github.com/apcera/gssapi/blob/5fb4217df13b8e6878046fe1e5c10e560e1b86dc/context.go#L310-L313

Which allows for release code with no nil checks:

https://github.com/openshift/origin/blob/d780740fac0ba7a9997be7dc35a985cba6e71b21/pkg/oc/util/tokencmd/negotiator_gssapi.go#L153-L155

Instead of the SSPI code that requires explicit nil checks:

https://github.com/enj/origin/blob/7d3bd2dd556c2962ab261a45d1b2521a880ede4e/pkg/oc/util/tokencmd/negotiator_sspi.go#L129-L134

> I don't see how it is possible for *ClientContext in normal user code to be nil. So if *ClientContext is nil by mistake, I want Release to panic, not to go away quietly.

Basically I have something that looks like:

```go

type state struct {
   // has credential and context
}

func (s *state) next(inToken) (outToken, done, error) {
     // do stuff to start, possibly error
     // build credential if needed, possibly error
     // build context if needed, possibly error
     // perform one step of the flow, possibly error
     // handles both starting and continuing a flow 
}

func (s *state) release() {
   // blindly release credential and context regardless of where next ended up
}
```

```go
s := &state{}
defer s.release()  // release everything at the end, no matter what
for !done {
    inToken := ...  // get token from HTTP header
    outToken, isDone, err := s.next(inToken)
    // do stuff with token
    // maybe set done
}
```

If you still disagree, no problem, I can drop the commit.  This was mostly a convenience change.

The real code for reference: https://github.com/openshift/origin/pull/11371

Reply:
Thank you for explaining. Fair enough. I will submit this as is.

Alex